### ISSUE TITLE
ModelResolutionException for incomplete java dependencies

### ISSUE THREAD

**Issue by sgalles** [ceylon/ceylon-compiler#2052]
To reproduce you need this Gradle jar **gradle-core-2.2.1.jar** from the Gradle 2.2.1 distribution (for convenience the jar is also available here http://stephane.galles.free.fr/ceylon/misc/tests/gradle-core-2.2.1.jar )

Import this jar in your Ceylon repos, create a Ceylon project, and import the jar in a module, for instance
```ceylon
module com.sgalles "1.0.0" {
     shared import gradlecore "2.2.1";
} 
```
Then in this project add this code :
```ceylon
import org.gradle.api {
    ...
}
shared class Foo(shared Project project){
    print(project.tasks.create("foo")); 
}     
```
Compiled with the CLI :
```
source\com\sgalles\run.ceylon:5: error: ExpressionVisitor caused an exception visiting QualifiedMemberExpression node: com.redh
at.ceylon.compiler.loader.ModelResolutionException: Failed to determine if findByPath is overriding a super method at com.redha
t.ceylon.compiler.java.loader.CeylonModelLoader.handleCompletionFailure(CeylonModelLoader.java:664)
    print(project.tasks.create("foo"));
                       ^
source\com\sgalles\run.ceylon:5: error: method or attribute does not exist: 'create' in type 'TaskContainer'
    print(project.tasks.create("foo"));
                       ^
2 errors
Note: Created module com.sgalles/1.0.0
ceylon compile: There were 2 errors
```
It seems that this error is related to missing dependencies (the complete Gradle distribution contains a bunch of other jars). If so, it would be more helpful to list the missing java classes in the CLI error message.
For instance Eclipse in a Java project can display messages sush as : 
```error
The type org.gradle.api.Action cannot be resolved. It is indirectly referenced from required .class files
```

**Comment by quintesse** [comment]
@sgalles If you import a JAR manually into your repositories you should use `ceylon import-jar` which would then advise you that you have a whole bunch of missing dependencies.

@FroMage The hidden causal error is `com.sun.tools.javac.code.Symbol$CompletionFailure: class file for org.gradle.util.Configurable not found` for which the message part is quite useful. But instead we're getting a compiler bug error (see `CeylonModelLoader` line 664. I tried looking at a better way of handling this but I'm not sure. Any ideas?

**Comment by sgalles** [comment]
@quintesse I used the `Export -> Java archive to Module repository` menu of Eclipse and I didn't get any error message AFAICT. Do you mean that the CLI `ceylon import-jar` will provide me with the list of missing dependencies ?

**Comment by gavinking** [comment]
Perhaps the IDE simply doesn't report the errors? It is supposed to be using the same code, isn't it?

**Comment by gavinking** [comment]
Ah no it doesn't the IDE is it's own thing.

**Comment by sgalles** [comment]
@quintesse @gavinking Ah yes ! With the CLI it is *very* clear (and very useful indeed) ! I was only relying on the IDE for this and I didn't know this hidden gem of the CLI...
```
Modules containing the following packages need to be declared as shared imports:
(Tip: try running again with the '--show-suggestions' option)
    ch.qos.logback.classic
    ch.qos.logback.classic.turbo
    ch.qos.logback.core
   [...]
    org.gradle.messaging.serialize
    org.gradle.model.entity.internal
    org.gradle.model.internal.core
    org.gradle.model.internal.inspect
    org.gradle.model.internal.registry
    org.gradle.util
    org.slf4j
```

**Comment by gavinking** [comment]
Well we need to somehow export this functionality so the IDE can reuse it.

**Comment by quintesse** [comment]
Yeah, I'm doing that now, but the API will be a bit horrible I'm afraid (the whole code is basically just about giving feedback to the user, making something that makes sense for the command line and is still  usable by the IDE does not really allow for a sane API). Well as long as it works I guess.

**Comment by tombentley** [comment]
So @quintesse are you working on this or not?

**Comment by quintesse** [comment]
I never worked on this @tombentley , what I did was what @gavinking suggested and I split off the importer code to be reused, see #2070

**Comment by tombentley** [comment]
So afaics the IDE hasn't been changed to use that new API, so I'll open an issue for that.

**Comment by tombentley** [comment]
Now the only thing remaining to do here is decide if we can/should do something better with the `CompletionFailure`. Looking at the code we're already chaining the exceptions. Given that:

* both the official ways to import jars into repos should now warn people about missing dependencies
* there's no easy way to distinguish an "expected" `CompletionFailure` from an unexpected one
* ultimately a `ModelResolutionException` is a fatal error for the code generator.

I suppose the only thing we could realistically improve would be to add the message of the `ModelResolutionException`'s cause to the error.

**Comment by gavinking** [comment]
> I never worked on this @tombentley , what I did was what @gavinking suggested and I split off the importer code to be reused, see #2070 

Well yeah, but it's useless to me in its current form.

**Comment by quintesse** [comment]
> it's useless to me in its current form.

Oh? Well I did ask for feedback but never heard from you so I thought everything was okay.

**Comment by gavinking** [comment]
You dumped it on me the day before you left for vacation.

**Comment by quintesse** [comment]
Which was already how long ago? Heh
But what's wrong with it?

**Comment by quintesse** [comment]
Maybe a better question is: what is it exactly that you need? What kind of API are you looking for?

**Comment by gavinking** [comment]
Well I don't really remember now. I wasted a couple of hours trying to get it to do something useful, and failed, and gave up. Whatever, it wasn't at all useful to the IDE.

**Comment by quintesse** [comment]
Well that's not really helpful for me to try and figure out how to change it. I'd at least have to know what kind of functionality you want.

**Comment by gavinking** [comment]
Well I'm trying to cast my mind back, not so easy to remember, but it's something like this:

What the Export wizard is supposed to do is:

1. Let the user select a jar to export.
2. Propose a list of needed module imports. (Currently not possible via the API.)
3. Let the user add to/remove from the list. (No obvious way to persist their choices via the API.)
4. Copy the jar to the target module repo.
5. Create a new module descriptor in the output target repo. (I was not able to convince the API to do this after trying many different ways. This might just be a bug, since there are some things in the API that look suspiciously like they _should_ make it possible to do this.)

So basically the only needed functionality this offered me was to copy a jar from one place to the another. Which is not really especially helpful.

It also offered me a very complex API for attaching some kind of warning logger, which wasn't going to be of much use to me in the IDE.
